owasm/chaintests/fair_crypto_market_price/src/lib.rs

```diff
@@ -0,0 +1,260 @@
+use borsh::{BorshDeserialize, BorshSchema, BorshSerialize};
+use owasm::{execute_entry_point, ext, oei, prepare_entry_point};
+
+#[derive(BorshDeserialize, BorshSchema)]
+struct Input {
+    base_symbol: String,
+    quote_symbol: String,
+    aggregation_method: String,
+    multiplier: u64,
+}
+
+#[derive(BorshSerialize, BorshSchema, Debug, PartialEq)]
+struct Output {
+    px: u64,
+}
+
+#[no_mangle]
+fn prepare_impl(input: Input) {
+    // Coingecko data source
+    oei::request_external_data(1, 10, &input.base_symbol.as_bytes());
+    oei::request_external_data(1, 11, &input.quote_symbol.as_bytes());
+
+    // Crypto compare source
+    oei::request_external_data(2, 20, &input.base_symbol.as_bytes());
+    oei::request_external_data(2, 21, &input.quote_symbol.as_bytes());
+
+    // Binance source
+    oei::request_external_data(3, 30, &input.base_symbol.as_bytes());
+    oei::request_external_data(3, 31, &input.quote_symbol.as_bytes());
+}
+
+fn len(arr: &Vec<f64>) -> f64 {
+    let mut l = 0f64;
+    for _ in arr.iter() {
+        l += 1f64
+    }
+    l
+}
+
+fn only_positive(arr: Vec<f64>) -> Vec<f64> {
+    arr.iter().filter(|&&x| x > 0f64).map(|&x| x).collect::<Vec<_>>()
+}
+
+fn mean(arr: Vec<f64>) -> f64 {
+    let pos_arr = only_positive(arr);
+    let len_pos_arr = len(&pos_arr);
+    if len_pos_arr > 0f64 {
+        pos_arr.iter().fold(0f64, |sum, val| sum + val) / len_pos_arr
+    } else {
+        0f64
+    }
+}
+
+fn median(arr: Vec<f64>) -> f64 {
+    let mut pos_arr = only_positive(arr);
+    let len_pos_arr = len(&pos_arr);
+    if len_pos_arr > 0f64 {
+        pos_arr.sort_by(|a, b| a.partial_cmp(b).unwrap());
+        let mid = len_pos_arr / 2f64;
+        if len_pos_arr as u64 % 2 == 0 {
+            let mut l = 0f64;
+            let mut x1 = 0f64;
+            let mut x2 = 0f64;
+            for &x in pos_arr.iter() {
```

Comment:
Can't we access array at mid-index instead of loop?

Reply:
done